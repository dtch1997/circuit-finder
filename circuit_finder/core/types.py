import torch
from typing import Protocol
from transformer_lens.hook_points import HookPoint

HookName = str
LayerIndex = int

ALL_GPT_2_SMALL_LAYERS: list[LayerIndex] = list(range(12))

device = "cuda" if torch.cuda.is_available() else "cpu"


class TransformerLensForwardHook(Protocol):
    def __call__(self, orig: torch.Tensor, hook: HookPoint) -> torch.Tensor:
        raise NotImplementedError


class TransformerLensBackwardHook(Protocol):
    def __call__(self, orig: torch.Tensor, hook: HookPoint) -> tuple[torch.Tensor]:
        raise NotImplementedError
