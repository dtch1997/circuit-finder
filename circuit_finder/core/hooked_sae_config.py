import transformer_lens as tl
from dataclasses import dataclass

@dataclass
class HookedSAEConfig(tl.HookedSAEConfig):
    # NOTE: When this is True, the error gradient is allowed to flow back to the input layer.
    # This is useful to do attribution patching for many nodes in parallel. 
    # Additionally, it's the approach taken in Marks et al. 
    allow_error_grad_to_input: bool = False 