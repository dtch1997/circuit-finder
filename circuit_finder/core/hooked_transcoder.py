import einops
import torch
import torch.nn as nn
import torch.nn.functional as F
import transformer_lens as tl
from jaxtyping import Float

from typing import Dict, Union, Sequence
from transformer_lens.hook_points import HookPoint, HookedRootModule

from circuit_finder.core.types import LayerIndex
from circuit_finder.core.hooked_transcoder_config import HookedTranscoderConfig


class HookedTranscoder(HookedRootModule):
    """Hooked Transcoder"""

    def __init__(self, cfg: Union[HookedTranscoderConfig, Dict]):
        super().__init__()
        if isinstance(cfg, Dict):
            cfg = HookedTranscoderConfig(**cfg)
        elif isinstance(cfg, str):
            raise ValueError(
                "Please pass in a config dictionary or HookedSAEConfig object."
            )
        self.cfg = cfg

        self.W_enc = nn.Parameter(
            torch.nn.init.kaiming_uniform_(
                torch.empty(self.cfg.d_in, self.cfg.d_sae, dtype=self.cfg.dtype)
            )
        )
        self.W_dec = nn.Parameter(
            torch.nn.init.kaiming_uniform_(
                torch.empty(self.cfg.d_sae, self.cfg.d_in, dtype=self.cfg.dtype)
            )
        )
        self.b_enc = nn.Parameter(torch.zeros(self.cfg.d_sae, dtype=self.cfg.dtype))
        self.b_dec = nn.Parameter(torch.zeros(self.cfg.d_in, dtype=self.cfg.dtype))

        self.hook_sae_input = HookPoint()
        self.hook_sae_acts_pre = HookPoint()
        self.hook_sae_acts_post = HookPoint()
        self.hook_sae_recons = HookPoint()

        self.to(self.cfg.device)
        self.setup()

    def maybe_reshape_input(
        self,
        input: Float[torch.Tensor, "... d_input"],
        apply_hooks: bool = True,
    ) -> Float[torch.Tensor, "... d_in"]:
        """
        Reshape the input to have correct dim.
        No-op for standard SAEs, but useful for hook_z SAEs.
        """
        if apply_hooks:
            self.hook_sae_input(input)

        if input.shape[-1] == self.cfg.d_in:
            x = input
        else:
            # Assume this this is an attention output (hook_z) SAE
            assert self.cfg.hook_name.endswith(
                "_z"
            ), f"You passed in an input shape {input.shape} does not match SAE input size {self.cfg.d_in} for hook_name {self.cfg.hook_name}. This is only supported for attn output (hook_z) SAEs."
            x = einops.rearrange(input, "... n_heads d_head -> ... (n_heads d_head)")
        assert (
            x.shape[-1] == self.cfg.d_in
        ), f"Input shape {x.shape} does not match SAE input size {self.cfg.d_in}"

        return x

    def encode(
        self,
        x: Float[torch.Tensor, "... d_in"],
        apply_hooks: bool = True,
    ) -> Float[torch.Tensor, "... d_sae"]:
        """SAE Encoder.

        Args:
            input: The input tensor of activations to the SAE. Shape [..., d_in].

        Returns:
            output: The encoded output tensor from the SAE. Shape [..., d_sae].
        """
        # Subtract bias term
        x_cent = x - self.b_dec

        # SAE hidden layer pre-RELU  activation
        sae_acts_pre = (
            einops.einsum(x_cent, self.W_enc, "... d_in, d_in d_sae -> ... d_sae")
            + self.b_enc  # [..., d_sae]
        )
        if apply_hooks:
            sae_acts_pre = self.hook_sae_acts_pre(sae_acts_pre)

        # SAE hidden layer post-RELU activation
        sae_acts_post = F.relu(sae_acts_pre)  # [..., d_sae]
        if apply_hooks:
            sae_acts_post = self.hook_sae_acts_post(sae_acts_post)

        return sae_acts_post

    def decode(
        self,
        sae_acts_post: Float[torch.Tensor, "... d_sae"],
        apply_hooks: bool = True,
    ) -> Float[torch.Tensor, "... d_in"]:
        x_reconstruct = (
            einops.einsum(
                sae_acts_post, self.W_dec, "... d_sae, d_sae d_in -> ... d_in"
            )
            + self.b_dec
        )
        if apply_hooks:
            x_reconstruct = self.hook_sae_recons(x_reconstruct)
        return x_reconstruct

    def maybe_reshape_output(
        self,
        input: Float[torch.Tensor, "... d_input"],
        output: Float[torch.Tensor, "... d_in"],
    ):
        output = output.reshape(input.shape)
        return output

    def forward(
        self, input: Float[torch.Tensor, "... d_in"]
    ) -> Float[torch.Tensor, "... d_in"]:
        """SAE Forward Pass.

        Args:
            input: The input tensor of activations to the SAE. Shape [..., d_in].
                Also supports hook_z activations of shape [..., n_heads, d_head], where n_heads * d_head = d_in, for attention output (hook_z) SAEs.

        Returns:
            output: The reconstructed output tensor from the SAE, with the error term optionally added. Same shape as input (eg [..., d_in])
        """
        x = self.maybe_reshape_input(input)
        sae_acts_post = self.encode(x)
        x_reconstruct = self.decode(sae_acts_post)
        return self.maybe_reshape_output(input, x_reconstruct)


class HookedTranscoderWrapper(HookedRootModule):
    """Wrapper around transcoder and the MLP it replaces"""

    def __init__(
        self,
        transcoder: HookedTranscoder,
        mlp: nn.Module,
    ):
        super().__init__()
        self.transcoder = transcoder

        self.hook_sae_error = HookPoint()
        self.hook_sae_output = HookPoint()
        self.setup()

        # NOTE: we want to exclude the MLP from the HookedTransformer's hook points.
        # So we add mlp after setup.
        # Suggested by Arthur
        self.mlp = mlp
        self.mlp.to(transcoder.cfg.device)

    @property
    def cfg(self):
        return self.transcoder.cfg

    def forward(self, x):
        sae_output = self.transcoder(x)
        if self.cfg.use_error_term:
            with torch.no_grad():
                clean_sae_out = self.transcoder(x)
                clean_mlp_out = self.mlp(x)
                sae_error = clean_mlp_out - clean_sae_out
            sae_error.requires_grad = True
            sae_error.retain_grad()
            sae_error = self.hook_sae_error(sae_error)
            sae_output += sae_error
        return self.hook_sae_output(sae_output)


def get_layer_of_hook_name(hook_point):
    return int(hook_point.split(".")[1])


class HookedTranscoderReplacementContext:
    """Context manager to replace MLP sublayers with transcoders"""

    model: tl.HookedTransformer
    transcoders: Sequence[HookedTranscoder]
    layers: Sequence[LayerIndex]
    original_mlps: dict[LayerIndex, nn.Module]

    def __init__(
        self, model: tl.HookedTransformer, transcoders: Sequence[HookedTranscoder]
    ):
        self.layers = [get_layer_of_hook_name(t.cfg.hook_name) for t in transcoders]
        self.original_mlps = {layer: model.blocks[layer].mlp for layer in self.layers}
        self.transcoders = transcoders
        self.model = model

    def __enter__(self):
        # Replace all MLPs with transcoder
        for layer, transcoder in zip(self.layers, self.transcoders):
            mlp = self.model.blocks[layer].mlp
            self.model.blocks[layer].mlp = HookedTranscoderWrapper(transcoder, mlp)

        self.model.setup()

        # # Replace original run_with_cache to include the transcoders' cache
        # self.orig_run_with_cache = self.model.run_with_cache

        # def run_with_cache(self: tl.HookedTransformer, *args, **kwargs):
        #     cache = self.model.run_with_cache(*args, **kwargs)

        # self.model.run_with_cache = run_with_cache

    def __exit__(self, exc_type, exc_value, exc_tb):
        # Restore original MLPs
        for layer, mlp in self.original_mlps.items():
            self.model.blocks[layer].mlp = mlp

        # # Restore original run_with_cache
        # self.model.run_with_cache = self.orig_run_with_cache
        self.model.setup()
