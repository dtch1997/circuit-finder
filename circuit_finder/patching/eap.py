"""Implement"""

import transformer_lens as tl
from circuit_finder.core.types import LayerIndex
from typing import Callable
from dataclasses import dataclass

from circuit_finder.core import HookedSAE

HookName = str
ModuleType = str
NodeType = str
sae_module_type = [
    "mlp",
    "mlp.transcoder",
    "hook_resid_pre",
    "hook_resid_post",
    "attn.hook_z",
]
sae_node_types = ["hook_sae_error", "hook_sae_output"]


@dataclass(frozen=True)
class Node:
    layer: LayerIndex
    module_type: ModuleType
    node_type: NodeType

    @property
    def hook_name(self):
        return f"blocks.{self.layer}.{self.module_type}.{self.node_type}"

    @staticmethod
    def from_hook_name(hook_name: str) -> "Node":
        parts = hook_name.split(".")
        layer = int(parts[1])
        node_type = parts[-1]
        module_type = ".".join(parts[2:-1])
        return Node(layer, module_type, node_type)


def get_gradient_hook_point_for_sae(sae: HookedSAE) -> HookName:
    """Get the gradient hook point for an SAE.

    Return a hook name in a standard HookedTransformer that will have the same gradient as sae_output.
    In practice this will usually be the residual stream.
    """
    sae_node = Node.from_hook_name(sae.cfg.hook_name)
    if sae_node.module_type in ("mlp", "mlp.transcoder"):  # blocks.{layer}.mlp.
        # Return hook resid post
        return f"blocks.{sae_node.layer}.hook_resid_post"
    elif sae_node.module_type == "attn":  # blocks.{layer}.attn.hook_z
        # Return hook resid mid
        return f"blocks.{sae_node.layer}.hook_resid_mid"
    elif sae_node.module_type in ("hook_resid_pre", "hook_resid_post"):
        # Here we can just return the same hook point
        return sae.cfg.hook_name

    raise ValueError(f"Unknown SAE module type {sae_node.module_type}")


class ForwardBackwardCache:
    """Context manager to cache forward and backward activations of a model"""

    def __init__(
        self, model: tl.HookedSAETransformer, filter_fn: str | Callable[[str], bool]
    ):
        self.model = model
        self.filter_fn = filter_fn
        self.act_cache = {}
        self.grad_cache = {}

    def __enter__(self):
        # TODO: change this to use ```with model.hooks(): ... '''
        self.model.reset_hooks()

        def forward_cache_hook(act, hook):
            self.act_cache[hook.name] = act.detach()

        def backward_cache_hook(act, hook):
            self.grad_cache[hook.name] = act.detach()

        self.model.add_hook(self.filter_fn, forward_cache_hook, "fwd")
        self.model.add_hook(self.filter_fn, backward_cache_hook, "bwd")
        return self

    def __exit__(self, exc_type, exc_value, exc_tb):
        self.model.reset_hooks()
