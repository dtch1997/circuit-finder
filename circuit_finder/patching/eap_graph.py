import pandas as pd
from typing import Callable
from circuit_finder.core.types import (
    LayerIndex,
    FeatureIndex,
    TokenIndex,
    Node,
    Edge,
    Attrib,
    ModuleName,
    parse_node_name,
)


def convert_graph_to_dataframe(graph: "EAPGraph") -> pd.DataFrame:
    rows = []
    for edge, edge_info in graph.graph:
        dest, src = edge
        if "metric" in src:
            continue
        nn_grad, nn_attrib, em_grad, em_attrib = edge_info
        src_module_type, src_layer, src_token, src_feature = parse_node_name(src)
        dest_module_type, dest_layer, dest_token, dest_feature = parse_node_name(dest)
        rows.append(
            {
                "src_module_type": src_module_type,
                "dest_module_type": dest_module_type,
                "src_layer": src_layer,
                "dest_layer": dest_layer,
                "src_token": src_token,
                "dest_token": dest_token,
                "src_feature": src_feature,
                "dest_feature": dest_feature,
                "nn_grad": nn_grad,
                "nn_attrib": nn_attrib,
                "em_grad": em_grad,
                "em_attrib": em_attrib,
            }
        )
    df = pd.DataFrame(rows)
    return df


class EAPGraph:
    """A class representing a circuit of nodes."""

    graph: list[tuple[Edge, Attrib]]

    def __init__(self, graph: list[tuple[Edge, Attrib]] = []):
        """Initialize the graph"""
        self.graph = graph

    def get_edges(
        self, filter_fn: Callable[[Edge], bool] = lambda x: True
    ) -> list[Edge]:
        """Get the edges of the graph"""
        return sorted([edge for edge, _ in self.graph if filter_fn(edge)])

    def get_nodes(
        self, filter_fn: Callable[[Node], bool] = lambda x: True
    ) -> list[Node]:
        """Get the nodes of the graph"""
        node_set = set()
        for dest, src in self.get_edges():
            if filter_fn(src):
                node_set.add(src)
            if filter_fn(dest):
                node_set.add(dest)
        return sorted(list(node_set))

    def get_src_nodes(
        self, filter_fn: Callable[[Node], bool] = lambda x: True
    ) -> list[Node]:
        """Get the source nodes of the graph"""
        node_set = set()
        for _, src in self.get_edges():
            if filter_fn(src):
                node_set.add(src)
        return sorted(list(node_set))

    def get_dest_nodes(
        self, filter_fn: Callable[[Node], bool] = lambda x: True
    ) -> list[Node]:
        """Get the destination nodes of the graph"""
        node_set = set()
        for dest, _ in self.get_edges():
            if filter_fn(dest):
                node_set.add(dest)
        return sorted(list(node_set))

    def to_json(self) -> dict:
        """Convert the graph to a JSON object"""
        return {
            "graph": [
                (str(dest), str(src), attrib) for (dest, src), attrib in self.graph
            ]
        }

    @staticmethod
    def from_json(json: dict) -> "EAPGraph":
        """Load the graph from a JSON object"""
        graph = [
            ((Node(dest), Node(src)), attrib) for dest, src, attrib in json["graph"]
        ]
        return EAPGraph(graph)

    def has_node(self, module_name, layer, feature) -> bool:
        """Check if a node is in the graph"""
        for node in self.get_nodes():
            _, layer_idx, _, feature_idx = parse_node_name(node)
            if layer_idx == layer and feature_idx == feature:
                return True
        return False


def get_feature_and_token_idx_of_nodes(
    graph: EAPGraph, module_name: ModuleName, layer_idx: LayerIndex
) -> tuple[list[FeatureIndex], list[TokenIndex]]:
    """Get the feature and token indices of upstream nodes for all nodes at a given module and layer

    Note: Only searches over upstream nodes in the graph.

    Returns:
        feature_indices : list of feature indices
        token_indices: list of token positions

    These can be zipped together to get the (token_idx, feature_idx) pairs of important nodes.

    Args:
        module_name : downstream module name.
        layer_idx : index of the downstream layer.
    """
    feature_indices: list[FeatureIndex] = []
    token_indices: list[TokenIndex] = []

    def filter_by_module_and_layer(node: Node) -> bool:
        (module, layer_idx, _, _) = parse_node_name(node)
        return module == module_name and layer_idx == layer_idx

    # Get all nodes that are currently in the graph
    nodes_set: set[Node] = set()
    for node in graph.get_src_nodes(filter_fn=filter_by_module_and_layer):
        nodes_set.add(node)
    nodes: list[Node] = list(nodes_set)

    # Filter by module and layer
    # TODO: It seems like we could do this previously but ig it doesn't matter.
    for node in nodes:
        (_, _, token_idx, feature_idx) = parse_node_name(node)
        feature_indices += [int(feature_idx)]
        token_indices += [int(token_idx)]
    return feature_indices, token_indices
