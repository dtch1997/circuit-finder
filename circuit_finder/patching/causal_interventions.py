#%%

import transformer_lens as tl
from functools import partial
from circuit_finder.pretrained import (
    load_attn_saes,
    load_resid_saes,
    load_hooked_mlp_transcoders,
)
from circuit_finder.core.hooked_transcoder import HookedTranscoderReplacementContext

# Initialize SAEs
attn_saes = load_attn_saes(use_error_term=True)
mlp_transcoders = load_hooked_mlp_transcoders(use_error_term=True)

# Load model
model = tl.HookedSAETransformer.from_pretrained("gpt2").cuda()




#%%
def ablation_hook(act, hook, id_to_ablate):
    assert hook.name.endswith("hook_sae_acts_post")
    act[:, :, id_to_ablate] = 0
    return act
#%%
def run_with_ablations(
        model, 
        attn_saes, 
        mlp_transcoders, 
        ablation_list
        ):
    
    # Reset hooks on all saes/transcoders
    for t in mlp_transcoders.values():
        t.reset_hooks()
    for a in attn_saes.values():
        a.reset_hooks()

    # Add ablation hooks to all saes/transcoders
    for module, layer, feature_id in ablation_list:
        temp_ablate_hook = partial(
            ablation_hook, 
            id_to_ablate=feature_id
            )
        
        if module == "mlp":     
            mlp_transcoders[layer].add_hook(
                "hook_sae_acts_post",
                temp_ablate_hook,
                "fwd"
            )

        elif module == "attn":
            attn_saes[layer].add_hook(
                "hook_sae_acts_post",
                temp_ablate_hook,
                "fwd"
            )

        else:
            print("modules must be attn or mlp")

    # Run model
    with HookedTranscoderReplacementContext(
        model,  # type: ignore
        transcoders=mlp_transcoders.values()
    ) as context:          

        with model.saes(saes=attn_saes.values()):
            logits  = model("Hi my name is Jacob")
    
    return logits


# %%
ablation_list = [("mlp", 10, id) for id in range(20)]
logits = run_with_ablations(model, 
                            attn_saes, 
                            mlp_transcoders, 
                            ablation_list)


