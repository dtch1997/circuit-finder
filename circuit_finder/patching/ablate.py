from einops import einsum, rearrange
import torch
from jaxtyping import Int, Float
from functools import partial
from circuit_finder.core.types import MetricFn, Tokens
from circuit_finder.patching.eap_graph import EAPGraph
from torch import Tensor
import transformer_lens as tl

from transcoders_slim.transcoder import Transcoder
from circuit_finder.pretrained.load_mlp_transcoders import ts_tc_to_hooked_tc
from circuit_finder.core.hooked_sae import HookedSAE
from circuit_finder.core.hooked_transcoder import HookedTranscoder
from circuit_finder.core.hooked_transcoder import HookedTranscoderReplacementContext


def get_mask(
    graph, module_name, layer, feature_acts: Float[Tensor, "batch seq n_features"]
):
    """
    Returns mask : [seq, n_features].
    mask[seq, feature_id] is False whenever {module_name}.{layer}.{seq}.{feature_id} is a node of graph.
    All other entries are True
    """
    nodes = graph.get_src_nodes()
    positions = [
        int(node.split(".")[2])
        for node in nodes
        if node.split(".")[:2] == [module_name, str(layer)]
    ]
    feature_ids = [
        int(node.split(".")[3])
        for node in nodes
        if node.split(".")[:2] == [module_name, str(layer)]
    ]
    mask = torch.ones_like(feature_acts[0]).bool()
    mask[positions, feature_ids] = False
    return mask


def get_metric_with_ablation(
    model: tl.HookedTransformer,
    graph: EAPGraph,
    tokens: Int[Tensor, "batch seq"],
    metric: MetricFn,
    transcoders,  # list
    attn_saes,  # list
    ablate_nodes: str | bool = "zero",  # options [False, "bm", "zero"]
    ablate_errors: str | bool = False,  # options [False, "bm", "zero"]
    first_ablated_layer: int = 2,  # Marks et al don't ablate first 2 layers
    freeze_attention: bool = False,
):
    model = add_ablation_hooks_to_model(
        model,
        graph,
        tokens,
        transcoders,
        attn_saes,
        ablate_nodes,
        ablate_errors,
        first_ablated_layer,
        freeze_attention,
    )
    return metric(model, tokens)


def get_forward_cache(
    model: tl.HookedSAETransformer,
    tokens: Tokens,
    transcoders: list[Transcoder] | list[HookedTranscoder],
    attn_saes: list[HookedSAE],
) -> tl.ActivationCache:
    hooked_transcoders = []
    for tc in transcoders:
        if isinstance(tc, Transcoder):
            tc = ts_tc_to_hooked_tc(tc)
        hooked_transcoders.append(tc)

    # Run model
    with HookedTranscoderReplacementContext(
        model,  # type: ignore
        transcoders=hooked_transcoders,
    ) as context:
        with model.saes(saes=attn_saes):  # type: ignore
            _, cache = model.run_with_cache(tokens)

    return cache


def add_ablation_hooks_to_model(
    model: tl.HookedTransformer,
    graph: EAPGraph,
    tokens: Int[Tensor, "batch seq"],
    transcoders,  # list
    attn_saes,  # list
    ablate_nodes: str | bool = "zero",  # options [False, "bm", "zero"]
    ablate_errors: str | bool = False,  # options [False, "bm", "zero"]
    first_ablated_layer: int = 2,  # Marks et al don't ablate first 2 layers
    freeze_attention: bool = False,
) -> tl.HookedTransformer:
    """Cache the activations of the model on a circuit, then return the model with relevant ablation hooks added"""
    assert ablate_errors in [False, "bm", "zero"]
    assert ablate_nodes in [False, "bm", "zero"]
    # Do clean FP to get batchmean (bm) feature acts, and reconstructions
    # No patching is done during this FP, so the hook fns return nothing
    # Note: MLP transcoders are a bit more fiddly than attn-SAEs, requiring us to cache at both mlp_in and mlp_out
    mlp_bm_feature_act_cache = {}
    attn_bm_feature_act_cache = {}
    mlp_recons_cache = {}
    mlp_error_cache = {}
    attn_error_cache = {}

    def mlp_in_cache_hook(act, hook, layer):
        assert hook.name.endswith("ln2.hook_normalized")
        t = transcoders[layer]
        feature_acts = torch.relu(
            einsum(
                act - t.b_dec,
                t.W_enc,
                "b s d_model, d_model n_features -> b s n_features",
            )
            + t.b_enc
        )
        recons = feature_acts @ t.W_dec + t.b_dec_out

        mlp_bm_feature_act_cache[layer] = feature_acts.mean(0)
        mlp_recons_cache[layer] = recons

    def mlp_out_cache_hook(act, hook, layer):
        assert hook.name.endswith("mlp_out")
        mlp_error_cache[layer] = act - mlp_recons_cache[layer]

    def attn_cache_hook(act, hook, layer):
        assert hook.name.endswith("hook_z")
        sae = attn_saes[layer]
        z_concat = rearrange(act, "b s n_heads d_head -> b s (n_heads d_head)")
        feature_acts = torch.relu(
            einsum(
                z_concat - sae.b_dec,
                sae.W_enc,
                "b s d_model, d_model n_features -> b s n_features",
            )
            + sae.b_enc
        )
        recons_concat = feature_acts @ sae.W_dec + sae.b_dec
        recons = rearrange(
            recons_concat,
            "b s (n_heads d_head) -> b s n_heads d_head",
            n_heads=model.cfg.n_heads,
        )

        attn_bm_feature_act_cache[layer] = feature_acts.mean(0)
        attn_error_cache[layer] = act - recons

    model.reset_hooks()
    for layer in range(model.cfg.n_layers):
        model.add_hook(
            f"blocks.{layer}.ln2.hook_normalized",
            partial(mlp_in_cache_hook, layer=layer),
            "fwd",
        )
        model.add_hook(
            f"blocks.{layer}.hook_mlp_out",
            partial(mlp_out_cache_hook, layer=layer),
            "fwd",
        )
        model.add_hook(
            f"blocks.{layer}.attn.hook_z", partial(attn_cache_hook, layer=layer), "fwd"
        )

    # Run forward pass and populate all caches
    _, pattern_cache = model.run_with_cache(
        tokens, return_type="loss", names_filter=lambda x: x.endswith("pattern")
    )
    assert len(pattern_cache) > 0
    assert len(mlp_bm_feature_act_cache) > 0
    assert len(attn_bm_feature_act_cache) > 0
    assert len(mlp_recons_cache) > 0
    assert len(mlp_error_cache) > 0
    assert len(attn_error_cache) > 0

    # Now do FP where we patch nodes not in graph with their batchmeaned values
    model.reset_hooks()
    mlp_ablated_recons = {}

    def freeze_pattern_hook(act, hook):
        assert hook.name.endswith("pattern")
        return pattern_cache[hook.name]

    def mlp_out_ablated_recons_cache_hook(act, hook, layer):
        assert hook.name.endswith("ln2.hook_normalized")
        t = transcoders[layer]
        feature_acts = torch.relu(
            einsum(
                act - t.b_dec,
                t.W_enc,
                "b s d_model, d_model n_features -> b s n_features",
            )
            + t.b_enc
        )
        mask = get_mask(graph, "mlp", layer, feature_acts)
        ablated_feature_acts = feature_acts.clone()
        ablated_feature_acts[:, mask] = mlp_bm_feature_act_cache[layer][mask]
        mlp_ablated_recons[layer] = ablated_feature_acts @ t.W_dec + t.b_dec_out

    def mlp_patch_hook(act, hook, layer):
        assert hook.name.endswith("mlp_out")

        if ablate_nodes == "bm":
            recons = mlp_ablated_recons[layer]
        elif ablate_nodes == "zero":
            recons = torch.zeros_like(mlp_ablated_recons[layer])
        else:
            # NOTE: This requires us to change the way we're computing the cached objects.
            raise NotImplementedError

        if ablate_errors == "bm":
            error = mlp_error_cache[layer].mean(dim=0, keepdim=True)
        elif ablate_errors == "zero":
            error = torch.zeros_like(mlp_error_cache[layer])
        else:
            error = mlp_error_cache[layer]

        return recons + error

    def attn_patch_hook(act, hook, layer):
        assert hook.name.endswith("hook_z")
        sae = attn_saes[layer]
        z_concat = rearrange(act, "b s n_heads d_head -> b s (n_heads d_head)")
        feature_acts = torch.relu(
            einsum(
                z_concat - sae.b_dec,
                sae.W_enc,
                "b s d_model, d_model n_features -> b s n_features",
            )
            + sae.b_enc
        )

        mask = get_mask(graph, "attn", layer, feature_acts)
        ablated_feature_acts = feature_acts.clone()
        ablated_feature_acts[:, mask] = attn_bm_feature_act_cache[layer][mask]
        ablated_recons_concat = ablated_feature_acts @ sae.W_dec + sae.b_dec
        ablated_recons = rearrange(
            ablated_recons_concat,
            "b s (n_heads d_head) -> b s n_heads d_head",
            n_heads=model.cfg.n_heads,
        )

        if ablate_nodes == "bm":
            recons = ablated_recons
        elif ablate_nodes == "zero":
            recons = torch.zeros_like(ablated_recons)
        else:  # no ablation
            raise NotImplementedError

        if ablate_errors == "bm":
            error = attn_error_cache[layer].mean(dim=0, keepdim=True)
        elif ablate_errors == "zero":
            error = torch.zeros_like(attn_error_cache[layer])
        else:
            error = attn_error_cache[layer]

        return recons + error

    for layer in range(first_ablated_layer, model.cfg.n_layers):
        model.add_hook(
            f"blocks.{layer}.ln2.hook_normalized",
            partial(mlp_out_ablated_recons_cache_hook, layer=layer),
            "fwd",
        )
        model.add_hook(
            f"blocks.{layer}.hook_mlp_out", partial(mlp_patch_hook, layer=layer), "fwd"
        )
        model.add_hook(
            f"blocks.{layer}.attn.hook_z", partial(attn_patch_hook, layer=layer), "fwd"
        )
        if freeze_attention:
            model.add_hook(
                f"blocks.{layer}.attn.hook_pattern", freeze_pattern_hook, "fwd"
            )

    return model
