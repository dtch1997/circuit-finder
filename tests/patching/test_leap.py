from circuit_finder.core.hooked_sae_config import HookedSAEConfig
from circuit_finder.core.hooked_sae import HookedSAE
from circuit_finder.core.hooked_transcoder import (
    HookedTranscoder,
)
from circuit_finder.core.hooked_transcoder_config import HookedTranscoderConfig
from circuit_finder.patching.leap import LEAP, LEAPConfig

import pytest
import torch

from transformer_lens import HookedSAETransformer

MODEL = "solu-2l"
prompt = "Hello World!"


class Counter:
    def __init__(self):
        self.count = 0

    def inc(self, *args, **kwargs):
        self.count += 1


@pytest.fixture(scope="module")
def model():
    model = HookedSAETransformer.from_pretrained(MODEL)
    yield model
    model.reset_saes()


def get_site(act_name):
    site = act_name.split(".")[2:]
    site = ".".join(site)
    return site


def get_transcoder_config(model, act_name_in, act_name_out):
    site_to_size = {
        "ln2.hook_normalized": model.cfg.d_model,
        "hook_mlp_out": model.cfg.d_model,
    }
    site_in = get_site(act_name_in)
    assert site_in == "ln2.hook_normalized"
    d_in = site_to_size[site_in]

    site_out = get_site(act_name_out)
    d_out = site_to_size[site_out]

    return HookedTranscoderConfig(
        d_in=d_in,
        d_out=d_out,
        d_sae=d_in * 2,
        hook_name=act_name_in,
        hook_name_out=act_name_out,
    )


def get_attn_sae_config(model, act_name):
    site_to_size = {
        "hook_z": model.cfg.d_head * model.cfg.n_heads,
        "hook_mlp_out": model.cfg.d_model,
        "hook_resid_pre": model.cfg.d_model,
        "hook_post": model.cfg.d_mlp,
    }
    site = act_name.split(".")[-1]
    d_in = site_to_size[site]
    return HookedSAEConfig(d_in=d_in, d_sae=d_in * 2, hook_name=act_name)


@pytest.mark.xfail(reason="Unsure why not determinstic")
def test_leap(model, snapshot):
    config = LEAPConfig(threshold=0.01, contrast_pairs=False, chained_attribs=True)
    tokens = model.to_tokens(prompt)

    # NOTE: All attn saes and transcoders must be hooked in
    # Seed the random generation
    torch.manual_seed(0)
    attn_saes = {
        layer: HookedSAE(get_attn_sae_config(model, f"blocks.{layer}.attn.hook_z"))
        for layer in range(model.cfg.n_layers)
    }

    transcoders = {
        layer: HookedTranscoder(
            get_transcoder_config(
                model,
                f"blocks.{layer}.ln2.hook_normalized",
                f"blocks.{layer}.hook_mlp_out",
            )
        )
        for layer in range(model.cfg.n_layers)
    }

    def metric_fn(model, tokens):
        return model(tokens, return_type="loss")

    leap = LEAP(
        config,
        tokens,
        model,
        attn_saes=attn_saes,
        transcoders=transcoders,
        metric=metric_fn,
    )

    leap.metric_step()
    for layer in reversed(range(1, model.cfg.n_layers)):
        leap.mlp_step(layer)
        leap.ov_step(layer)

    graph = leap.graph
    assert graph == snapshot
