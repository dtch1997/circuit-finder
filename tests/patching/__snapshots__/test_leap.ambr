# serializer version: 1
# name: test_leap
  None
# ---
